x-pack/plugins/security_solution/server/lib/siem_migrations/rules/api/util/installation.ts

```diff
@@ -31,105 +30,100 @@ const installPrebuiltRules = async (
-    (value, index, self) => index === self.findIndex((rule) => rule.rule_id === value.rule_id)
   );
 
+  // Install prebuilt rules
   // TODO: we need to do an error handling which can happen during the rule installation
-  const { results: installedRules } = await createPrebuiltRules(
+  const { results: newlyInstalledRules } = await createPrebuiltRules(
     detectionRulesClient,
-    filteredAssetsToInstall
+    installable
   );
   await performTimelinesInstallation(securitySolutionContext);
 
+  const installedRules = [
+    ...alreadyInstalledRules,
+    ...newlyInstalledRules.map((value) => value.result),
+  ];
+
+  // Create migration rules updates templates
+  const rulesToUpdate: UpdateRuleMigrationInput[] = [];
   installedRules.forEach((installedRule) => {
-    const rules = rulesToInstall.filter(
-      (rule) => rule.elastic_rule?.prebuilt_rule_id === installedRule.result.rule_id
+    const filteredRules = rulesToInstall.filter(
+      (rule) => rule.elastic_rule?.prebuilt_rule_id === installedRule.rule_id
     );
-    rules.forEach((prebuiltRule) => {
-      rulesToUpdate.push({
-        id: prebuiltRule.id,
+    rulesToUpdate.push(
+      ...filteredRules.map(({ id }) => ({
+        id,
         elastic_rule: {
-          id: installedRule.result.id,
+          id: installedRule.id,
         },
-      });
-    });
+      }))
+    );
   });
 
   return rulesToUpdate;
 };
 
-const installCustomRules = async (
+export const installCustomRules = async (
   rulesToInstall: StoredRuleMigration[],
   detectionRulesClient: IDetectionRulesClient,
   logger: Logger
 ): Promise<UpdateRuleMigrationInput[]> => {
   const rulesToUpdate: UpdateRuleMigrationInput[] = [];
-  await Promise.all(
-    rulesToInstall.map(async (rule) => {
+  const createCustomRulesOutcome = await initPromisePool({
+    concurrency: MAX_RULES_TO_UPDATE_IN_PARALLEL,
+    items: rulesToInstall,
+    executor: async (rule) => {
       if (!rule.elastic_rule?.query || !rule.elastic_rule?.description) {
         return;
       }
-      try {
-        const payloadRule: RuleCreateProps = {
-          type: 'esql',
-          language: 'esql',
-          query: rule.elastic_rule.query,
-          name: rule.elastic_rule.title,
-          description: rule.elastic_rule.description,
-          severity: DEFAULT_TRANSLATION_SEVERITY,
-          risk_score: DEFAULT_TRANSLATION_RISK_SCORE,
-        };
-        const createdRule = await detectionRulesClient.createCustomRule({
-          params: payloadRule,
-        });
-        rulesToUpdate.push({
-          id: rule.id,
-          elastic_rule: {
-            id: createdRule.id,
-          },
-        });
-      } catch (err) {
-        // TODO: we need to do an error handling which can happen during the rule creation
-        logger.debug(`Could not create a rule because of error: ${JSON.stringify(err)}`);
-      }
-    })
-  );
+      const payloadRule: RuleCreateProps = {
+        type: 'esql',
+        language: 'esql',
```

Comment:
we have `rule.elastic_rule.query_language`

Reply:
Yes and these two fields are both required for rule creation and need to be passed. That is why I'm passing them both as `esql`. Or do you mean something else?

Reply:
I mean that maybe we should use `elastic_rule.query_language` instead of hardcoding `'esql'`.

Reply:
ah yeah, we can do that!